pkg/distributor/distributor.go

```diff
@@ -415,12 +535,46 @@ func (d *Distributor) sendSamplesErr(ctx context.Context, ingester ring.Instance
 
 	_, err = c.(logproto.PusherClient).Push(ctx, req)
 	d.ingesterAppends.WithLabelValues(ingester.Addr).Inc()
-	if err != nil {
-		d.ingesterAppendFailures.WithLabelValues(ingester.Addr).Inc()
+
+	if err == nil {
+		return nil
+	}
+
+	d.ingesterAppendFailures.WithLabelValues(ingester.Addr).Inc()
+
+	// TODO: look for a better way to recognize a per-stream rate limit.
+	if d.cfg.ShardStreams.Mode == NeverShardMode {
+		return err
+	}
+
+	if errMsg := err.Error(); strings.Contains(errMsg, "Per stream rate limit exceeded (limit") {
+		d.streamSharding.Inc()
+		shardLimitedStream(errMsg, d.streamSharder)
 	}
 	return err
 }
 
+// shardLimitedStream shards limited stream based on the given per-stream rate limited message.
+func shardLimitedStream(perStreamErrorMsg string, sharder StreamSharder) {
+	re := regexp.MustCompile("\\{(.*?)\\}")
+	matches := re.FindAllStringSubmatch(perStreamErrorMsg, 1)
+	if len(matches) < 1 {
+		return
+	}
+
+	labels := fmt.Sprintf("{%s}", matches[0][1]) // add curly braces
```

Comment:
If we can avoid `fmt.Sprintf` I think we can save a bit of CPU (`Sprintf` uses reflection) and possibly an alloc; the equivalent will work:
```suggestion
	labels := "{" + matches[0][1] + "}"
```